api/internal/plugins/loader/loader.go

```diff
@@ -95,13 +100,47 @@ func relativePluginPath(id resid.ResId) string {
 		strings.ToLower(id.Kind))
 }
 
-func AbsolutePluginPath(pc *types.PluginConfig, id resid.ResId) string {
-	return filepath.Join(
-		pc.AbsPluginHome, relativePluginPath(id), id.Kind)
+func (l *Loader) AbsolutePluginPath(id resid.ResId) (string, error) {
+	pluginHome, err := l.absPluginHome()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(pluginHome, relativePluginPath(id), id.Kind), nil
 }
 
-func (l *Loader) absolutePluginPath(id resid.ResId) string {
-	return AbsolutePluginPath(l.pc, id)
+// absPluginHome is the home of kustomize Exec and Go plugins.
+// Kustomize plugin configuration files are k8s-style objects
+// containing the fields 'apiVersion' and 'kind', e.g.
+//   apiVersion: apps/v1
+//   kind: Deployment
+// kustomize reads plugin configuration data from a file path
+// specified in the 'generators:' or 'transformers:' field of a
+// kustomization file.  For Exec and Go plugins, kustomize
+// uses this data to both locate the plugin and configure it.
+// Each Exec or Go plugin (its code, its tests, its supporting data
+// files, etc.) must be housed in its own directory at
+//   ${absPluginHome}/${pluginApiVersion}/LOWERCASE(${pluginKind})
+// where
+//   - ${absPluginHome} is an absolute path, defined below.
+//   - ${pluginApiVersion} is taken from the plugin config file.
+//   - ${pluginKind} is taken from the plugin config file.
+func (l *Loader) absPluginHome() (string, error) {
+	// External plugins are disabled--return the dummy plugin root.
+	if l.pc.PluginRestrictions != types.PluginRestrictionsNone {
+		return konfig.NoPluginHomeSentinal, nil
+	}
+	// We've already determined plugin home--use the cached value.
+	if l.absolutePluginHome != "" {
+		return l.absolutePluginHome, nil
+	}
+
+	// Check default locations for a valid plugin root, and cache it if found.
+	dir, err := konfig.DefaultAbsPluginHome(filesys.MakeFsOnDisk())
```

Comment:
Using a real file system down deep this way inhibits testing with an inmemory fs.

Fine with adding an injected `filesys.FileSystem` member to PluginConfig

It's a struggle to keep this code simple.  One of the purposes of the various loaders is to encapsulate what can be legally loaded.

Reply:
not dead set on adding all this to pluginconfig, just posing q's

Also fine with adding a FileSystem field to the plugin loader.

I'd suggest adding a fileloader 
https://github.com/kubernetes-sigs/kustomize/blob/master/api/loader/fileloader.go
to the plugin loader, but it's restrictions would have to be turned off or modified for it to read from some system default plugin location.   By default it cannot read from an absolute path directory unless that directory has a kustomization file in it.  That prevents someone from, say, reading /etc/passwd into a configmap.

So there are two levels of file abstraction - the file system abstraction (in mem vs on disk), and the loader abstraction which imposes security rules by examining paths.

Reply:
> not dead set on adding all this to pluginconfig, just posing q's

> Also fine with adding a FileSystem field to the plugin loader

I'm going to push a commit with this soon since I had so much imports trouble with the PluginConfig suggestion (I can still do that if you want, but I'll need a bit more guidance).

I'm not sure I understand the fileloader suggestion though--given that the environment variables that configure plugin roots can point anywhere the user chooses, what will the fileloader's restrictions give us in this case? Is enhancing this required for this PR? I'm still using the `konfig.NoPluginHomeSentinal` technique that was in place before, but now it gets used by `absPluginHome` above regardless of the field value, so in a way it is slightly stronger?